Source/Urho3D/CSharp/Managed/Container/SharedPtr.cs

```diff
@@ -0,0 +1,146 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace Urho3DNet
+{
+    public class SharedPtr<T> : IDisposable, IEquatable<SharedPtr<T>> where T : RefCounted
+    {
+        /// <summary>
+        /// Construct immutable SharedPtr object.
+        /// Don't forget to dispose it.
+        /// </summary>
+        /// <param name="other">Object to store pointer to.</param>
+        public SharedPtr(T other)
+        {
+            if (other?.IsNotExpired == true)
+            {
+                _value = other;
+                _value.AddRef();
+            }
+            else
+                _value = null;
+        }
+
+        /// <summary>
+        /// SharedPtr finalizer.
+        /// Should not be executed on not-disposed SharedPtr.
+        /// </summary>
+        ~SharedPtr()
+        {
+            if (_value?.IsNotExpired == true)
+                Trace.WriteLine(
+                    $"Object of type {typeof(T).FullName} wasn't properly disposed. Please call Dispose on all SharedPtr<{typeof(T).Name}> instances.");
+        }
+
+        /// <summary>
+        /// Copy SharedPtr.
+        /// </summary>
+        /// <param name="other">SharedPtr instance to copy</param>
+        public SharedPtr(SharedPtr<T> other) : this(other?._value)
+        {
+        }
+
+        /// <summary>Indicates whether the current object is equal to another object of the same type.</summary>
+        /// <param name="other">An object to compare with this object.</param>
+        /// <returns>true if the current object is equal to the <paramref name="other">other</paramref> parameter; otherwise, false.</returns>
+        public bool Equals(SharedPtr<T> other)
+        {
+            if (ReferenceEquals(null, other)) return false;
+            if (ReferenceEquals(this, other)) return true;
+            return EqualityComparer<T>.Default.Equals(_value, other._value);
+        }
+
+        /// <summary>Determines whether the specified object is equal to the current object.</summary>
+        /// <param name="obj">The object to compare with the current object.</param>
+        /// <returns>true if the specified object  is equal to the current object; otherwise, false.</returns>
+        public override bool Equals(object obj)
+        {
+            if (ReferenceEquals(null, obj)) return false;
+            if (ReferenceEquals(this, obj)) return true;
+            if (obj.GetType() != this.GetType()) return false;
+            return Equals((SharedPtr<T>)obj);
+        }
+
+        /// <summary>Serves as the default hash function.</summary>
+        /// <returns>A hash code for the current object.</returns>
+        public override int GetHashCode()
+        {
+            return EqualityComparer<T>.Default.GetHashCode(_value);
```

Comment:
Maybe we should also use `T.getCPtr(_value).Handle` for hashing as it feels safer. In theory we should not have two different managed instances pointing to same `RefCounted`, so this is probably fine.. But using raw C++ pointer feels more explicit and thus appropriate.

Reply:
No reason to waste time on equality methods now if they aren't in use. I'm going to delete this.